micro_dl/utils/mp_utils.py

```diff
@@ -47,53 +48,54 @@ def create_save_mask(input_fnames,
     :param int int2str_len: Length of str when converting ints
     :param str mask_type: thresholding type used for masking or str to map to
      masking function
+    :param str mask_ext: 'npy' or 'png'. Save the mask as uint8 PNG or
+         NPY files
     :return dict cur_meta for each mask
     """
 
-    im_stack = np.squeeze(tile_utils.read_imstack(
+    im_stack = tile_utils.read_imstack(
         input_fnames,
         flat_field_fname,
         normalize_im=False,
-    ))
-    # Combine channel images and generate mask
-    if len(im_stack.shape) == 2:
-        summed_image = im_stack
-    elif len(im_stack.shape) == 3:
-        if len(input_fnames) == 1:
-            # read a 3d image
-            summed_image = im_stack
-        else:
-            # read a 2d image stack
-            summed_image = np.sum(np.stack(im_stack), axis=2)
-    elif len(im_stack.shape) == 4:
-        # read a 3d image stack
-        summed_image = np.sum(np.stack(im_stack), axis=3)
-    summed_image = summed_image.astype('float32')
-    if mask_type == 'otsu':
-        mask = mask_utils.create_otsu_mask(summed_image, str_elem_radius)
-    elif mask_type == 'unimodal':
-        mask = mask_utils.create_unimodal_mask(summed_image,
-                                               str_elem_radius)
+    )
+    masks = []
+    for idx in range(im_stack.shape[-1]):
+        im = im_stack[..., idx].astype('float32')
+        if mask_type == 'otsu':
+            mask = mask_utils.create_otsu_mask(im, str_elem_radius)
+        elif mask_type == 'unimodal':
+            mask = mask_utils.create_unimodal_mask(im,  str_elem_radius)
+        masks += [mask]
+    masks = np.stack(masks, axis=-1)
+    mask = np.any(masks, axis=-1)
 
     # Create mask name for given slice, time and position
     file_name = aux_utils.get_im_name(time_idx=time_idx,
                                       channel_idx=mask_channel_idx,
                                       slice_idx=slice_idx,
                                       pos_idx=pos_idx,
                                       int2str_len=int2str_len)
-    # Save mask for given channels, mask is 2D
-    np.save(os.path.join(mask_dir, file_name),
-            mask,
-            allow_pickle=True,
-            fix_imports=True)
+
+    if mask_ext == 'npy':
+        # Save mask for given channels, mask is 2D
+        np.save(os.path.join(mask_dir, file_name),
+                mask,
+                allow_pickle=True,
+                fix_imports=True)
+    elif mask_ext == 'png':
+        file_name = file_name[:-3] + 'png'
+        # Covert mask to uint8
+        mask = mask.astype(np.uint8) * (2 ** 8 - 1)
```

Comment:
I think it's fine to just write 255. Or if you want to be fancy you can use np.iinfo

Reply:
Wondering how this will be read with openCV, will the unique values be 0 and 255. That might an issue here: https://github.com/czbiohub/microDL/blob/876085b13060650ce955a202cf09506f120974ad/micro_dl/utils/tile_utils.py#L179 as the mean is not a volume fraction or percent anymore!

Reply:
Good point. I think it should be fine if the images are converted to bool after reading them, e.g. here: https://github.com/czbiohub/microDL/blob/876085b13060650ce955a202cf09506f120974ad/micro_dl/utils/tile_utils.py#L29